### PR TITLE
Data parallel helper

### PR DESCRIPTION
**Pull request by angeloskath**
This PR adds float16 and bfloat16 types to MPI as well as a helper to average gradients over a distributed group. It is a bit ugly but it helps nicely on https://github.com/ml-explore/mlx-examples/pull/821 . So I am not sure if it belongs in `nn.utils` 🤔 .

Here is a simple scaling plot for a Llama 8B finetuning.

llama-8b-finetuning

### CODE REVIEW

File: python/mlx/nn/utils.py [hunk 1]
```diff
@@ -68,3 +69,93 @@ def wrapped_checkpointed_fn(*args, **kwargs):
...
+    def _average(x):
+        dt = x.dtype
+        x = x.astype(communication_type) if communication_type is not None else x
+        return mx.distributed.all_sum(x, stream=mx.cpu).astype(dt) / N
+
+    if all_reduce_size <= 0:
+        return tree_map(_average, gradients)
+
+    else:
+        flat_grads = sorted(tree_flatten(gradients), key=lambda x: x[0])
```
Comment (awni) [inline python/mlx/nn/utils.py:113]: What is the purpose of the sort here?

File: python/mlx/nn/utils.py [hunk 2]
```diff
@@ -68,3 +69,93 @@ def wrapped_checkpointed_fn(*args, **kwargs):
         return checkpointed_fn(module.trainable_parameters(), *args, **kwargs)
 
     return wrapped_checkpointed_fn
+
+
+def average_gradients(
```
Comment (awni) [inline python/mlx/nn/utils.py:74]: I'm not sure it matters, but this function has two different return modes which could lead to some confusion:
- Return reference to input tree if it's a no-op
- Return copied tree structure o/w

So if you do:

```
avg_grads = average_gradients(grads)
grads[0] = ...
```

The behavior would be different in the two cases. It seems like a really odd usage pattern.. but nevertheless it might be worth making a copy of the tree structure in all cases.

**Review by awni — APPROVED** [review]
Looks really nice. I think of all the places this could be `nn.utils` is probably the best. If you think there is a world in which we want to do more distributed stuff in `nn`.. we could make a new sub-package `nn.distributed` and put it there (which may also be a good home for the distributed layers in #1270).

**Comment by awni** [comment]
The benchmark scaling is pretty remarkable btw.. so much potential for fast distributed fine-tuning!

File: python/mlx/nn/utils.py (hunk 1 above)
Comment (angeloskath) [reply python/mlx/nn/utils.py:113]: Yeah this should be either removed or the code just above is wrong. Basically in order for this to work `tree_map` needs to have a deterministic ordering across machines (because `all_sum` needs to be called on the equivalent arrays).

So this comes down to iteration ordering of dicts in python. Looking into it a bit more since python 3.7 the iteration is guaranteed to be the order of insertion which is why the code in line 110 works as well. So the conclusion is I will remove the sort :-)

File: python/mlx/nn/utils.py (hunk 2 above)
Comment (angeloskath) [reply python/mlx/nn/utils.py:74]: Hm hadn't thought of that. One reason to keep it as is would be that in the case of no-op we have absolutely no overhead (except a python function call which is nanoseconds). Otherwise doing a copy would add some overhead proportional to the number of parameters.

Let me know what you think. It could be fine as it is likely hidden behind computation anyway.

Comment (awni) [reply python/mlx/nn/utils.py:74]: Yea I don't like making overhead for no reason and those trees can get pretty large. So probably let's just keep it the way it is and deal with it in the future if needed.

File: python/mlx/nn/utils.py (hunk 1 above)
Comment (awni) [reply python/mlx/nn/utils.py:113]: I hadn't thought of the ordering across machines. That makes sense!
